File: ios/Plugin/CacheManager.swift
import Foundation

/**
 * Cache Manager for iOS
 * 缓存管理器 - iOS实现
 */
@objc public class CacheManager: NSObject {
    
    // MARK: - Types
    
    /**
     * Cache Entry
     */
    @objc public class CacheEntry: NSObject {
        @objc public var key: String
        @objc public var data: String
        @objc public var headers: [String: String]
        @objc public var status: Int
        @objc public var timestamp: TimeInterval
        @objc public var expiresAt: TimeInterval
        @objc public var size: Int
        @objc public var compressed: Bool
        @objc public var metadata: CacheMetadata?
        
        init(key: String, data: String, headers: [String: String], status: Int, maxAge: TimeInterval) {
            self.key = key
            self.data = data
            self.headers = headers
            self.status = status
            self.timestamp = Date().timeIntervalSince1970
            self.expiresAt = self.timestamp + maxAge
            self.size = data.utf8.count
            self.compressed = false
            super.init()
        }
        
        @objc public func isExpired() -> Bool {
            return Date().timeIntervalSince1970 > expiresAt
        }
        
        @objc public func toDictionary() -> [String: Any] {
            var dict: [String: Any] = [
                "key": key,
                "data": data,
                "headers": headers,
                "status": status,
                "timestamp": timestamp,
                "expiresAt": expiresAt,
                "size": size,
                "compressed": compressed
            ]
            
            if let metadata = metadata {
                dict["metadata"] = metadata.toDictionary()
            }
            
            return dict
        }
        
        @objc public static func fromDictionary(_ dict: [String: Any]) -> CacheEntry? {
            guard let key = dict["key"] as? String,
                  let data = dict["data"] as? String,
                  let headers = dict["headers"] as? [String: String],
                  let status = dict["status"] as? Int,
                  let timestamp = dict["timestamp"] as? TimeInterval,
                  let expiresAt = dict["expiresAt"] as? TimeInterval else {
                return nil
            }
            
            let entry = CacheEntry(key: key, data: data, headers: headers, status: status, maxAge: 0)
            entry.timestamp = timestamp
            entry.expiresAt = expiresAt
            entry.size = dict["size"] as? Int ?? data.utf8.count
            entry.compressed = dict["compressed"] as? Bool ?? false
            
            if let metadataDict = dict["metadata"] as? [String: Any] {
                entry.metadata = CacheMetadata.fromDictionary(metadataDict)
            }
            
            return entry
        }
    }
    
    /**
     * Cache Metadata
     */
    @objc public class CacheMetadata: NSObject {
        @objc public var url: String
        @objc public var method: String
        @objc public var etag: String?
        @objc public var lastModified: String?
        
        init(url: String, method: String) {
            self.url = url
            self.method = method
            super.init()
        }
        
        func toDictionary() -> [String: Any] {
            var dict: [String: Any] = [
                "url": url,
                "method": method
            ]
            if let etag = etag { dict["etag"] = etag }
            if let lastModified = lastModified { dict["lastModified"] = lastModified }
            return dict
        }
        
        static func fromDictionary(_ dict: [String: Any]) -> CacheMetadata? {
            guard let url = dict["url"] as? String,
                  let method = dict["method"] as? String else {
                return nil
            }
            
            let metadata = CacheMetadata(url: url, method: method)
            metadata.etag = dict["etag"] as? String
            metadata.lastModified = dict["lastModified"] as? String
            return metadata
        }
    }
    
    /**
     * Cache Statistics
     */
    @objc public class CacheStats: NSObject {
        @objc public var totalEntries: Int = 0
        @objc public var totalSize: Int = 0
        @objc public var hits: Int = 0
        @objc public var misses: Int = 0
        @objc public var hitRate: Double = 0.0
        @objc public var oldestEntry: TimeInterval = 0
        @objc public var newestEntry: TimeInterval = 0
        
        func toDictionary() -> [String: Any] {
            return [
                "totalEntries": totalEntries,
                "totalSize": totalSize,
                "hits": hits,
                "misses": misses,
                "hitRate": hitRate,
                "oldestEntry": oldestEntry,
                "newestEntry": newestEntry
            ]
        }
    }
    
    /**
     * Cache Eviction Policy
     */
    @objc public enum CacheEvictionPolicy: Int {
        case lru  // Least Recently Used
        case lfu  // Least Frequently Used
        case fifo // First In First Out
        case ttl  // Time To Live
    }
    
    // MARK: - Properties
    
    private var memoryCache: [String: CacheEntry] = [:]
    private var accessCount: [String: Int] = [:]
    private var accessTime: [String: TimeInterval] = [:]
    private let lock = NSLock()
    private let maxSize: Int
    private let evictionPolicy: CacheEvictionPolicy
    private let cacheDirectory: URL
    
    private var currentSize: Int = 0
    private var hits: Int = 0
    private var misses: Int = 0
    
    // MARK: - Initialization
    
    @objc public init(maxSize: Int, evictionPolicy: CacheEvictionPolicy) {
        self.maxSize = maxSize
        self.evictionPolicy = evictionPolicy
        
        // Setup cache directory
        let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        self.cacheDirectory = cacheDir.appendingPathComponent("http_cache")
        
        super.init()
        
        // Create cache directory if needed
        try? FileManager.default.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        
        // Calculate initial cache size
        calculateCacheSize()
        
        print("[CacheManager] Initialized with maxSize: \(maxSize), policy: \(evictionPolicy)")
    }
    
    // MARK: - Public Methods
    
    /**
     * Get cache entry
     */
    @objc public func get(_ key: String) -> CacheEntry? {
        lock.lock()
        defer { lock.unlock() }
        
        // Check memory cache first
        var entry = memoryCache[key]
        
        if entry == nil {
            // Try to load from disk
            entry = loadFromDisk(key)
            if let entry = entry {
                memoryCache[key] = entry
            }
        }
        
        guard let entry = entry else {
            misses += 1
            return nil
        }
        
        // Check if expired
        if entry.isExpired() {
            delete(key)
            misses += 1
            return nil
        }
        
        // Update access statistics
        accessCount[key] = (accessCount[key] ?? 0) + 1
        accessTime[key] = Date().timeIntervalSince1970
        hits += 1
        
        return entry
    }
    
    /**
     * Put cache entry
     */
    @objc public func put(_ key: String, entry: CacheEntry) {
        lock.lock()
        defer { lock.unlock() }
        
        // Check if we need to evict
        if currentSize + entry.size > maxSize {
            evict(requiredSpace: entry.size)
        }
        
        // Add to memory cache
        memoryCache[key] = entry
        accessCount[key] = 1
        accessTime[key] = Date().timeIntervalSince1970
        
        // Save to disk
        saveToDisk(key, entry: entry)
        
        currentSize += entry.size
        
        print("[CacheManager] Cached: \(key) (size: \(entry.size))")
    }
    
    /**
     * Delete cache entry
     */
    @objc public func delete(_ key: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        
        if let entry = memoryCache.removeValue(forKey: key) {
            currentSize -= entry.size
        }
        
        accessCount.removeValue(forKey: key)
        accessTime.removeValue(forKey: key)
        
        // Delete from disk
        let fileURL = cacheDirectory.appendingPathComponent(getCacheFileName(key))
        try? FileManager.default.removeItem(at: fileURL)
        
        return true
    }
    
    /**
     * Clear all cache
     */
    @objc public func clear() {
        lock.lock()
        defer { lock.unlock() }
        
        memoryCache.removeAll()
        accessCount.removeAll()
        accessTime.removeAll()
        
        // Delete all files
        if let files = try? FileManager.default.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: nil) {
            for file in files {
                try? FileManager.default.removeItem(at: file)
            }
        }
        
        currentSize = 0
        hits = 0
        misses = 0
        
        print("[CacheManager] Cache cleared")
    }
    
    /**
     * Check if key exists
     */
    @objc public func has(_ key: String) -> Bool {
        return get(key) != nil
    }
    
    /**
     * Get all cache keys
     */
    @objc public func keys() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        
        return Array(memoryCache.keys)
    }
    
    /**
     * Get cache statistics
     */
    @objc public func getStats() -> CacheStats {
        lock.lock()
        defer { lock.unlock() }
        
        let stats = CacheStats()
        stats.totalEntries = memoryCache.count
        stats.totalSize = currentSize
        stats.hits = hits
        stats.misses = misses
        stats.hitRate = (hits + misses) > 0 ? Double(hits) / Double(hits + misses) : 0
        
        var oldest = TimeInterval.greatestFiniteMagnitude
        var newest: TimeInterval = 0
        
        for entry in memoryCache.values {
            if entry.timestamp < oldest { oldest = entry.timestamp }
            if entry.timestamp > newest { newest = entry.timestamp }
        }
        
        stats.oldestEntry = oldest != TimeInterval.greatestFiniteMagnitude ? oldest : 0
        stats.newestEntry = newest
        
        return stats
    }
    
    /**
     * Cleanup expired entries
     */
    @objc public func cleanup() -> Int {
        lock.lock()
        defer { lock.unlock() }
        
        var cleaned = 0
        var toRemove: [String] = []
        
        for (key, entry) in memoryCache {
            if entry.isExpired() {
                toRemove.append(key)
            }
        }
        
        for key in toRemove {
            _ = delete(key)
            cleaned += 1
        }
        
        print("[CacheManager] Cleaned up \(cleaned) expired entries")
        return cleaned
    }
    
    /**
     * Get current cache size
     */
    @objc public func getSize() -> Int {
        lock.lock()
        defer { lock.unlock() }
        
        return currentSize
    }
    
    // MARK: - Private Methods
    
    /**
     * Evict entries to free space
     */
    private func evict(requiredSpace: Int) {
        let targetSize = Int(Double(maxSize) * 0.7) // Evict to 70%
        var freedSpace = 0
        
        var entries = Array(memoryCache)
        
        // Sort based on eviction policy
        switch evictionPolicy {
        case .lru:
            entries.sort { (a, b) in
                let timeA = accessTime[a.key] ?? 0
                let timeB = accessTime[b.key] ?? 0
                return timeA < timeB
            }
        case .lfu:
            entries.sort { (a, b) in
                let countA = accessCount[a.key] ?? 0
                let countB = accessCount[b.key] ?? 0
                return countA < countB
            }
        case .fifo:
            entries.sort { $0.value.timestamp < $1.value.timestamp }
        case .ttl:
            entries.sort { $0.value.expiresAt < $1.value.expiresAt }
        }
        
        for (key, entry) in entries {
            if currentSize - freedSpace <= targetSize && freedSpace >= requiredSpace {
                break
            }
            
            freedSpace += entry.size
            _ = delete(key)
        }
        
        print("[CacheManager] Evicted entries, freed: \(freedSpace) bytes")
    }
    
    /**
     * Save entry to disk
     */
    private func saveToDisk(_ key: String, entry: CacheEntry) {
        let fileURL = cacheDirectory.appendingPathComponent(getCacheFileName(key))
        
        do {
            let data = try JSONSerialization.data(withJSONObject: entry.toDictionary())
            try data.write(to: fileURL)
        } catch {
            print("[CacheManager] Error saving to disk: \(key), error: \(error)")
        }
    }
    
    /**
     * Load entry from disk
     */
    private func loadFromDisk(_ key: String) -> CacheEntry? {
        let fileURL = cacheDirectory.appendingPathComponent(getCacheFileName(key))
        
        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            return nil
        }
        
        do {
            let data = try Data(contentsOf: fileURL)
            let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            return dict.flatMap { CacheEntry.fromDictionary($0) }
        } catch {
            print("[CacheManager] Error loading from disk: \(key), error: \(error)")
            return nil
        }
    }
    
    /**
     * Get cache file name
     */
    private func getCacheFileName(_ key: String) -> String {
        return "\(key.hashValue).cache"
    }
    
    /**
     * Calculate total cache size
     */
    private func calculateCacheSize() {
        currentSize = 0
        
        if let files = try? FileManager.default.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: [.fileSizeKey]) {
            for file in files {
                if let size = try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize {
                    currentSize += size
                }
            }
        }
    }
}

File: android/src/main/java/com/capacitor/cors/CacheManager.java
package com.capacitor.cors;

import android.content.Context;
import android.util.Log;
import com.getcapacitor.JSObject;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Cache Manager for Android
 * 缓存管理器 - Android实现
 */
public class CacheManager {
    private static final String TAG = "CacheManager";
    private static final String CACHE_DIR = "http_cache";
    
    private final Context context;
    private final ConcurrentHashMap<String, CacheEntry> memoryCache;
    private final File cacheDir;
    private final long maxSize;
    private final CacheEvictionPolicy evictionPolicy;
    
    private long currentSize = 0;
    private int hits = 0;
    private int misses = 0;

    /**
     * Cache Entry
     */
    public static class CacheEntry {
        public String key;
        public String data;
        public Map<String, String> headers;
        public int status;
        public long timestamp;
        public long expiresAt;
        public long size;
        public boolean compressed;
        public CacheMetadata metadata;
        
        public CacheEntry(String key, String data, Map<String, String> headers, 
                         int status, long maxAge) {
            this.key = key;
            this.data = data;
            this.headers = headers;
            this.status = status;
            this.timestamp = System.currentTimeMillis();
            this.expiresAt = this.timestamp + maxAge;
            this.size = data.getBytes().length;
            this.compressed = false;
        }
        
        public boolean isExpired() {
            return System.currentTimeMillis() > expiresAt;
        }
        
        public JSONObject toJSON() throws JSONException {
            JSONObject json = new JSONObject();
            json.put("key", key);
            json.put("data", data);
            json.put("status", status);
            json.put("timestamp", timestamp);
            json.put("expiresAt", expiresAt);
            json.put("size", size);
            json.put("compressed", compressed);
            
            JSONObject headersJson = new JSONObject();
            for (Map.Entry<String, String> entry : headers.entrySet()) {
                headersJson.put(entry.getKey(), entry.getValue());
            }
            json.put("headers", headersJson);
            
            if (metadata != null) {
                json.put("metadata", metadata.toJSON());
            }
            
            return json;
        }
        
        public static CacheEntry fromJSON(JSONObject json) throws JSONException {
            String key = json.getString("key");
            String data = json.getString("data");
            int status = json.getInt("status");
            long timestamp = json.getLong("timestamp");
            long expiresAt = json.getLong("expiresAt");
            
            Map<String, String> headers = new HashMap<>();
            JSONObject headersJson = json.getJSONObject("headers");
            Iterator<String> keys = headersJson.keys();
            while (keys.hasNext()) {
                String headerKey = keys.next();
                headers.put(headerKey, headersJson.getString(headerKey));
            }
            
            CacheEntry entry = new CacheEntry(key, data, headers, status, 0);
            entry.timestamp = timestamp;
            entry.expiresAt = expiresAt;
            entry.size = json.optLong("size", data.getBytes().length);
            entry.compressed = json.optBoolean("compressed", false);
            
            if (json.has("metadata")) {
                entry.metadata = CacheMetadata.fromJSON(json.getJSONObject("metadata"));
            }
            
            return entry;
        }
    }
    
    /**
     * Cache Metadata
     */
    public static class CacheMetadata {
        public String url;
        public String method;
        public String etag;
        public String lastModified;
        
        public JSONObject toJSON() throws JSONException {
            JSONObject json = new JSONObject();
            json.put("url", url);
            json.put("method", method);
            if (etag != null) json.put("etag", etag);
            if (lastModified != null) json.put("lastModified", lastModified);
            return json;
        }
        
        public static CacheMetadata fromJSON(JSONObject json) throws JSONException {
            CacheMetadata metadata = new CacheMetadata();
            metadata.url = json.getString("url");
            metadata.method = json.getString("method");
            metadata.etag = json.optString("etag", null);
            metadata.lastModified = json.optString("lastModified", null);
            return metadata;
        }
    }
    
    /**
     * Cache Statistics
     */
    public static class CacheStats {
        public int totalEntries;
        public long totalSize;
        public int hits;
        public int misses;
        public double hitRate;
        public long oldestEntry;
        public long newestEntry;
        
        public JSObject toJSObject() {
            JSObject obj = new JSObject();
            obj.put("totalEntries", totalEntries);
            obj.put("totalSize", totalSize);
            obj.put("hits", hits);
            obj.put("misses", misses);
            obj.put("hitRate", hitRate);
            obj.put("oldestEntry", oldestEntry);
            obj.put("newestEntry", newestEntry);
            return obj;
        }
    }
    
    /**
     * Cache Eviction Policy
     */
    public enum CacheEvictionPolicy {
        LRU,  // Least Recently Used
        LFU,  // Least Frequently Used
        FIFO, // First In First Out
        TTL   // Time To Live
    }
    
    /**
     * Constructor
     */
    public CacheManager(Context context, long maxSize, CacheEvictionPolicy policy) {
        this.context = context;
        this.maxSize = maxSize;
        this.evictionPolicy = policy != null ? policy : CacheEvictionPolicy.LRU;
        this.memoryCache = new ConcurrentHashMap<>();
        this.cacheDir = new File(context.getCacheDir(), CACHE_DIR);
        
        if (!cacheDir.exists()) {
            cacheDir.mkdirs();
        }
        
        // Load cache size
        calculateCacheSize();
        
        Log.i(TAG, "CacheManager initialized with maxSize: " + maxSize + 
                   ", policy: " + this.evictionPolicy);
    }
    
    /**
     * Get cache entry
     */
    public CacheEntry get(String key) {
        // Check memory cache first
        CacheEntry entry = memoryCache.get(key);
        
        if (entry == null) {
            // Try to load from disk
            entry = loadFromDisk(key);
            if (entry != null) {
                memoryCache.put(key, entry);
            }
        }
        
        if (entry == null) {
            misses++;
            return null;
        }
        
        // Check if expired
        if (entry.isExpired()) {
            delete(key);
            misses++;
            return null;
        }
        
        hits++;
        return entry;
    }
    
    /**
     * Put cache entry
     */
    public void put(String key, CacheEntry entry) {
        // Check if we need to evict
        if (currentSize + entry.size > maxSize) {
            evict(entry.size);
        }
        
        // Add to memory cache
        memoryCache.put(key, entry);
        
        // Save to disk
        saveToDisk(key, entry);
        
        currentSize += entry.size;
        
        Log.d(TAG, "Cached: " + key + " (size: " + entry.size + ")");
    }
    
    /**
     * Delete cache entry
     */
    public boolean delete(String key) {
        CacheEntry entry = memoryCache.remove(key);
        
        if (entry != null) {
            currentSize -= entry.size;
        }
        
        // Delete from disk
        File file = new File(cacheDir, getCacheFileName(key));
        if (file.exists()) {
            return file.delete();
        }
        
        return entry != null;
    }
    
    /**
     * Clear all cache
     */
    public void clear() {
        memoryCache.clear();
        
        // Delete all files
        File[] files = cacheDir.listFiles();
        if (files != null) {
            for (File file : files) {
                file.delete();
            }
        }
        
        currentSize = 0;
        hits = 0;
        misses = 0;
        
        Log.i(TAG, "Cache cleared");
    }
    
    /**
     * Check if key exists
     */
    public boolean has(String key) {
        CacheEntry entry = get(key);
        return entry != null;
    }
    
    /**
     * Get all cache keys
     */
    public List<String> keys() {
        return new ArrayList<>(memoryCache.keySet());
    }
    
    /**
     * Get cache statistics
     */
    public CacheStats getStats() {
        CacheStats stats = new CacheStats();
        stats.totalEntries = memoryCache.size();
        stats.totalSize = currentSize;
        stats.hits = hits;
        stats.misses = misses;
        stats.hitRate = (hits + misses) > 0 ? (double) hits / (hits + misses) : 0;
        
        long oldest = Long.MAX_VALUE;
        long newest = 0;
        
        for (CacheEntry entry : memoryCache.values()) {
            if (entry.timestamp < oldest) oldest = entry.timestamp;
            if (entry.timestamp > newest) newest = entry.timestamp;
        }
        
        stats.oldestEntry = oldest != Long.MAX_VALUE ? oldest : 0;
        stats.newestEntry = newest;
        
        return stats;
    }
    
    /**
     * Cleanup expired entries
     */
    public int cleanup() {
        int cleaned = 0;
        List<String> toRemove = new ArrayList<>();
        
        for (Map.Entry<String, CacheEntry> entry : memoryCache.entrySet()) {
            if (entry.getValue().isExpired()) {
                toRemove.add(entry.getKey());
            }
        }
        
        for (String key : toRemove) {
            delete(key);
            cleaned++;
        }
        
        Log.i(TAG, "Cleaned up " + cleaned + " expired entries");
        return cleaned;
    }
    
    /**
     * Get current cache size
     */
    public long getSize() {
        return currentSize;
    }
    
    /**
     * Evict entries to free space
     */
    private void evict(long requiredSpace) {
        long targetSize = (long) (maxSize * 0.7); // Evict to 70%
        long freedSpace = 0;
        
        List<Map.Entry<String, CacheEntry>> entries = new ArrayList<>(memoryCache.entrySet());
        
        // Sort based on eviction policy
        switch (evictionPolicy) {
            case LRU:
                entries.sort((a, b) -> Long.compare(a.getValue().timestamp, b.getValue().timestamp));
                break;
            case FIFO:
                entries.sort((a, b) -> Long.compare(a.getValue().timestamp, b.getValue().timestamp));
                break;
            case TTL:
                entries.sort((a, b) -> Long.compare(a.getValue().expiresAt, b.getValue().expiresAt));
                break;
        }
        
        for (Map.Entry<String, CacheEntry> entry : entries) {
            if (currentSize - freedSpace <= targetSize && freedSpace >= requiredSpace) {
                break;
            }
            
            freedSpace += entry.getValue().size;
            delete(entry.getKey());
        }
        
        Log.i(TAG, "Evicted entries, freed: " + freedSpace + " bytes");
    }
    
    /**
     * Save entry to disk
     */
    private void saveToDisk(String key, CacheEntry entry) {
        try {
            File file = new File(cacheDir, getCacheFileName(key));
            FileOutputStream fos = new FileOutputStream(file);
            fos.write(entry.toJSON().toString().getBytes());
            fos.close();
        } catch (Exception e) {
            Log.e(TAG, "Error saving to disk: " + key, e);
        }
    }
    
    /**
     * Load entry from disk
     */
    private CacheEntry loadFromDisk(String key) {
        try {
            File file = new File(cacheDir, getCacheFileName(key));
            if (!file.exists()) {
                return null;
            }
            
            FileInputStream fis = new FileInputStream(file);
            byte[] data = new byte[(int) file.length()];
            fis.read(data);
            fis.close();
            
            String json = new String(data);
            return CacheEntry.fromJSON(new JSONObject(json));
        } catch (Exception e) {
            Log.e(TAG, "Error loading from disk: " + key, e);
            return null;
        }
    }
    
    /**
     * Get cache file name
     */
    private String getCacheFileName(String key) {
        return String.valueOf(key.hashCode()) + ".cache";
    }
    
    /**
     * Calculate total cache size
     */
    private void calculateCacheSize() {
        currentSize = 0;
        File[] files = cacheDir.listFiles();
        if (files != null) {
            for (File file : files) {
                currentSize += file.length();
            }
        }
    }
}